test/cpp/api/autograd.cpp

```diff
@@ -432,6 +432,79 @@ TEST(CustomAutogradTest, ReentrantPriority) {
   ASSERT_EQ(order.back(), 0);
 }
 
+TEST(CustomAutogradTest, Hooks) {
+  Variable x = torch::ones({5,5}, torch::requires_grad());
+  Variable y = torch::ones({5,5})*4;
+  y.set_requires_grad(true);
+
+  int counter = 0;
+
+  std::function<void(int, Variable)> bw_hook([&counter](int inc, Variable grad){
+    counter += inc;
+  });
+
+  Variable z = x * x + x * 2 + x * y + y;
+  x.register_hook(std::bind(bw_hook, 0 ,std::placeholders::_1));
```

Comment:
Unless necessary (e.g., moving arguments into a lambda), it's generally better not to use `std::bind`, and just pass in a lambda that does the binding.